system/jlib/jhash.ipp

```diff
@@ -255,37 +255,67 @@ class HashKeyOf : public MappingBase
 };
 
 
+template <class MAPPING>
+class KeptHashTableBase : extends KeptHashTable
+{
+  public:
+    typedef KeptHashTableBase<MAPPING> SELF;
+    friend class ConstHashItem;
+    class ConstHashItem
```

Comment:
since realised these need to be public classes so begin() and end() can be delegated.  Will need to go back to the other examples to make them consistent.